### PR TITLE
aseprite: 1.3.6 -> 1.3.7

### PR DESCRIPTION
**Pull request by ghost**
## Description of changes

* Update version
* Update dependencies: `tinyxml` -> `tinyxml-2`, `pcre` -> `pcre2`
* Add patch to use the shared `fmt` library
* Add translations
* Use clang as it's recommended and better supported by upstream (and seems to compile faster): https://github.com/aseprite/aseprite/blob/main/INSTALL.md#gcc-compiler
* Move to `pkgs/by-name`
* Move `skia-aseprite` to separate toplevel package
* Format with `nixfmt`
* Remove some non-idiomatic code

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by jtbx — APPROVED** [review]

**Comment by pbsds** [comment]
```
error: hash mismatch in fixed-output derivation '/nix/store/z99v2pjxv92y0bm27wiyrhqqycqwjwdv-source.drv':
         specified: sha256-75kYJXmyags0cW2D5Ksq1uUrFSCAkFOdmn7Ya/6jLXc=
            got:    sha256-vbPrH2MXU95DAKv5w64N5vdsBPpeNu657VrXCwhUDf8=
```

**Comment by vigress8** [comment]
@pbsds oops. Try now

File: pkgs/applications/editors/aseprite/default.nix [hunk 1]
```diff
@@ -12,42 +12,62 @@
...
+    # Translation strings
+    strings = fetchFromGitHub {
+      owner = "aseprite";
+      repo = "strings";
+      rev = "e18a09fefbb6cd904e506183d5fbe08558a52ed4";
+      hash = "sha256-GyCCxbhgf0vST20EH/+KkNLrF+U9Xzgpxlao8s925PQ=";
+    };
+  in ''
     sed -i src/ver/CMakeLists.txt -e "s-set(VERSION \".*\")-set(VERSION \"$version\")-"
+    rm -rf data/strings && cp -r ${strings} $_
```
Comment (pbsds) [inline pkgs/applications/editors/aseprite/default.nix:64]:
```suggestion
    rm -rf data/strings
    cp -r ${strings} data/strings
```

**Review by pbsds — COMMENTED** [review]
```
aseprite> patching file third_party/CMakeLists.txt
aseprite> patch unexpectedly ends in middle of line
aseprite> Hunk #1 succeeded at 33 with fuzz 2 (offset 1 line).
aseprite> applying patch /nix/store/7qzr3v9bsb85mnbyi2ndlps84f923lqp-shared-skia-deps.patch
aseprite> can't find file to patch at input line 3
aseprite> Perhaps you used the wrong -p or --strip option?
aseprite> The text leading up to this was:
aseprite> --------------------------
aseprite> |--- src/laf/cmake/FindSkia.cmake.orig	2022-01-08 02:15:13.417619266 +0100
aseprite> |+++ src/laf/cmake/FindSkia.cmake	2022-01-08 02:15:43.603960491 +0100
aseprite> --------------------------
```

**Comment by pbsds** [comment]
Would you be willing to adopt this package?

**Comment by vigress8** [comment]
> I am very impressed with your changes, would you be willing to adopt this package?

Sure!

**Comment by pbsds** [comment]
Then please add yourself in [maintainer-list.nix](https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix) with a commit titled `maintainers: add <user>` which preceeds the version bump, then add yourself to `meta.maintainers` and squash the package commits

**Comment by superherointj** [comment]
Suggestions:
- use nixfmt formatter (named `nixfmt-rfc-style` in unstable)
- move package to `pkgs/by-name/as/aseprite/` (requires removing reference from `pkgs/top-level/all-packages.nix`)

**Comment by vigress8** [comment]
I actually made most of those changes at first but didn't want to produce a huge diff. But if I have your OK for it then I'll move forward with those.
Also, it seems that aseprite's `skia` isn't in the binary cache, is it? If so, what'd be the way to enable caching for it?

File: pkgs/by-name/as/aseprite/package.nix [hunk 2]
```diff
@@ -12,42 +12,60 @@
 let
   skia = callPackage ./skia.nix {};
 in
-stdenv.mkDerivation rec {
+clangStdenv.mkDerivation (finalAttrs: {
```
Comment (superherointj) [inline pkgs/by-name/as/aseprite/package.nix:40]: Why to change `stdenv` to `clangStdenv`?

**Comment by superherointj** [comment]
Result of `nixpkgs-review pr 317258` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>aseprite</li>
  </ul>
</details>

**Comment by superherointj** [comment]
> I actually made most of those changes at first but didn't want to produce a huge diff. But if I have your OK for it then I'll move forward with those. 

Please go ahead with changes I listed. Such as, ordering by phases order is the usual way to organize the derivation.

> Also, it seems that aseprite's `skia` isn't in the binary cache, is it? If so, what'd be the best way to enable caching for it? Make it a separate package?

Packages from top-level are cached by default. Likely Skia needs to be packaged in separate (and shared by many packages). Several packages/applications use Skia, but it seems, each one uses it's own skia version. I don't know why this needs to be like that and can't be packaged globally.

File: pkgs/by-name/as/aseprite/package.nix (hunk 2 above)
Comment (vigress8) [reply pkgs/by-name/as/aseprite/package.nix:40]: Upstream recommends against compiling with gcc: https://github.com/aseprite/aseprite/blob/main/INSTALL.md#gcc-compiler
So I thought it's better to use clang by default

**Comment by vigress8** [comment]
> Is it failing build for you too?

Seems this is because I disabled `fetchSubmodules`, and the patch is applied to a submodule. But just enabling `fetchSubmodules` clones a lot of unnecessary stuff that isn't used. Is it possible to fetch only that submodule?

**Comment by superherointj** [comment]
> > Is it failing build for you too?
> 
> Seems this is because I disabled `fetchSubmodules`, and the patch is applied to a submodule. But just enabling `fetchSubmodules` clones a lot of unnecessary stuff that isn't used. Is it possible to fetch only that submodule?

Disabling `fetchSubmodules` is good. Check if dependency doesn't exist already in nixpkgs. On existence, if necessary, override patch there. If doesn't exist, it's possible to package it as a Nix package.

Also, consider adding `skia` to `passthru`. Such that, you can build/test in isolation.

**Comment by vigress8** [comment]
I think the cleanest way forward is prob to make `skia-aseprite` and `laf` into separate packages. Can I do it all in this PR or is it more idiomatic to open separate PRs?

**Comment by superherointj** [comment]
> I think the cleanest way forward is to make `skia-aseprite` and `laf` into separate top-level packages. Can I do it all in this PR or is it more idiomatic to open separate PRs?

Feel free to do it in this PR, in different commits though.
On commit order, order dependencies first, downstream packages at last.

**Comment by vigress8** [comment]
Turns out there's way too much vendoring going on, so I'll leave the submodules alone for now. It's doable to make them behave but too tedious

**Comment by superherointj** [comment]
> Turns out there's way too much vendoring going on, so I'll leave the submodules alone for now. It's doable to make them behave but too tedious

No problem. You have made a lot of progress here already.

File: pkgs/by-name/as/aseprite/package.nix [hunk 3]
```diff
@@ -106,7 +124,7 @@ stdenv.mkDerivation rec {
         This version is not redistributable: https://dev.aseprite.org/2016/09/01/new-source-code-license/
         Consider supporting the developer: https://aseprite.org/#buy
       '';
-    maintainers = with maintainers; [ orivej ];
-    platforms = platforms.linux;
+    maintainers = with lib.maintainers; [ orivej vigress8 ];
```
Comment (superherointj) [inline pkgs/by-name/as/aseprite/package.nix]: It is ideal to make a different commit for adding yourself as maintainer, the reason for that is that it can be reversed directly whenever you decide to no longer be a committer. And it won't affect the package.

File: pkgs/by-name/sk/skia-aseprite/package.nix [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    zlib
+  ];
+
+  buildPhase = ''
+    runHook preBuild
+    ninja -C lib skia modules
+    runHook postBuild
+  '';
+
+  installPhase = ''
```
Comment (superherointj) [inline pkgs/by-name/sk/skia-aseprite/package.nix:80]: install phase hooks are missing.

File: pkgs/by-name/sk/skia-aseprite/package.nix [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+      src/gpu/**/*.h \
+      third_party/externals/angle2/include \
+      third_party/skcms/**/*.h
+  '';
+
+  meta = {
+    description = "Complete 2D graphic library for drawing Text, Geometries, and Images (Aseprite's fork)";
+    homepage = "https://skia.org/";
+    license = lib.licenses.bsd3;
+    maintainers = with lib.maintainers; [ ];
```
Comment (superherointj) [inline pkgs/by-name/sk/skia-aseprite/package.nix]: I suggest you link to `aseprite.meta.maintainers` here. Because it is a dependency of it.

**Comment by superherointj** [comment]
CI errors as:

> pkgs/by-name/as/aseprite: File package.nix at line 92 contains the path expression "./shared-skia-deps.patch" which cannot be resolved: No such file or directory (os error 2)

File: pkgs/by-name/sk/skia-aseprite/package.nix [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+let
+  # deps.nix is generated by:
+  # ./update-deps.sh angle2 dng_sdk piex sfntly
+  depSrcs = import ./deps.nix { inherit fetchgit; };
+in
+clangStdenv.mkDerivation (finalAttrs: {
+  pname = "skia-aseprite";
+  version = "m102-861e4743af";
+
+  passthru.updateScript = gitUpdater { };
```
Comment (superherointj) [inline pkgs/by-name/sk/skia-aseprite/package.nix:32]: Please move `passthru` to just before `meta`.

File: pkgs/by-name/sk/skia-aseprite/package.nix [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+      modules/skshaper/include/*.h \
+      src/core/*.h \
+      src/gpu/**/*.h \
+      third_party/externals/angle2/include \
+      third_party/skcms/**/*.h
+  '';
+
+  meta = {
+    description = "Complete 2D graphic library for drawing Text, Geometries, and Images (Aseprite's fork)";
+    homepage = "https://skia.org/";
```
Comment (superherointj) [inline pkgs/by-name/sk/skia-aseprite/package.nix:106]: > downloadPage = "https://github.com/aseprite/skia";

**Comment by superherointj** [comment]
I've noticed Skia has only 8 commits ahead of upstream. It's possible to attempt to use upstream Skia and apply these 8 patches over it. I haven't looked into it. But this can unify Skia if it works.

**Comment by superherointj** [comment]
nixpkgs-review errors as:

> pkgs/by-name/as/aseprite/shared-skia-deps.patch': No such file or directory

File: pkgs/by-name/as/aseprite/package.nix [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+# Unfree version is not redistributable:
+# https://dev.aseprite.org/2016/09/01/new-source-code-license/
+# Consider supporting the developer: https://aseprite.org/#buy
+
+clangStdenv.mkDerivation (finalAttrs: {
+  pname = "aseprite";
+  version = "1.3.7";
+
+  passthru.updateScript = gitUpdater { rev-prefix = "v"; };
```
Comment (superherointj) [inline pkgs/by-name/as/aseprite/package.nix:42]:
```suggestion
  passthru.updateScript = gitUpdater { rev-prefix = "v"; };
```
Move `passthru.updateScript` to just before meta.

**Comment by vigress8** [comment]
Yeah sorry, the patch error is cause I forgot to add a file

File: pkgs/by-name/as/aseprite/package.nix [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+                - Export/import animations to/from Sprite Sheets.
+                - Tiled drawing mode, useful to draw patterns and textures.
+                - Undo/Redo for every operation.
+                - Real-time animation preview.
+                - Multiple editors support.
+                - Pixel-art specific tools like filled Contour, Polygon, Shading mode, etc.
+                - Onion skinning.
+
+              This version is not redistributable: https://dev.aseprite.org/2016/09/01/new-source-code-license/
+              Consider supporting the developer: https://aseprite.org/#buy
```
Comment (superherointj) [inline pkgs/by-name/as/aseprite/package.nix:169]:
```suggestion
```

Is this not redistributable? Then it's a problem, no?

Comment (vigress8) [reply pkgs/by-name/as/aseprite/package.nix:169]: The binary isn't redistributable, but compiling the source code for personal purposes is allowed. I guess that needs to be clarified

**Comment by vigress8** [comment]
Can the `update-deps.sh` script for `skia-aseprite` be made to run together with the `gitUpdater`?

**Comment by superherointj** [comment]
> Can the `update-deps.sh` script for `skia-aseprite` be made to run together with the `gitUpdater`?

I don't know the perfect solution. If you make a script triggering them both I expect it to work.

**Comment by vigress8** [comment]
All suggestions applied

File: pkgs/by-name/as/aseprite/package.nix [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+                - Export/import animations to/from Sprite Sheets.
+                - Tiled drawing mode, useful to draw patterns and textures.
+                - Undo/Redo for every operation.
+                - Real-time animation preview.
+                - Multiple editors support.
+                - Pixel-art specific tools like filled Contour, Polygon, Shading mode, etc.
+                - Onion skinning.
+
+              The binary files are not redistributable: https://dev.aseprite.org/2016/09/01/new-source-code-license/
+              Consider supporting the developer: https://aseprite.org/#buy
```
Comment (superherointj) [inline pkgs/by-name/as/aseprite/package.nix:169]:
```suggestion
```

File: pkgs/by-name/as/aseprite/package.nix [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+  pkg-config,
+  skia-aseprite,
+  tinyxml-2,
+  zlib,
+}:
+
+# The binary files are not redistributable:
+# https://dev.aseprite.org/2016/09/01/new-source-code-license/
+# Consider supporting the developer: https://aseprite.org/#buy
+
```
Comment (superherointj) [inline pkgs/by-name/as/aseprite/package.nix:37]:
```suggestion

```

File: pkgs/by-name/as/aseprite/package.nix [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+      dst="$out"/share/icons/hicolor/"$size"x"$size"
+      install -Dm644 "$src"/ase"$size".png "$dst"/apps/aseprite.png
+      install -Dm644 "$src"/doc"$size".png "$dst"/mimetypes/image-x-aseprite.png
+    done
+    # Delete unneeded artifacts of bundled libraries.
+    rm -rf "$out"/{include,lib,man}
+  '';
+
+  passthru.updateScript = gitUpdater { rev-prefix = "v"; };
+  meta = {
```
Comment (superherointj) [inline pkgs/by-name/as/aseprite/package.nix:150]:
```suggestion
  passthru.updateScript = gitUpdater { rev-prefix = "v"; };

  meta = {
```

File: pkgs/by-name/sk/skia-aseprite/package.nix [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+      src/core/*.h \
+      src/gpu/**/*.h \
+      third_party/externals/angle2/include \
+      third_party/skcms/**/*.h
+
+    runHook postInstall
+  '';
+
+  passthru.updateScript = gitUpdater { };
+  meta = {
```
Comment (superherointj) [inline pkgs/by-name/sk/skia-aseprite/package.nix:108]:
```suggestion
  passthru.updateScript = gitUpdater { };

  meta = {
```

**Comment by superherointj** [comment]
> Can the `update-deps.sh` script for `skia-aseprite` be made to run together with the `gitUpdater`?

I have not tested. I suggest to use bash to execute the two different routines. For fetching dependencies.

This currently errors:
> nix-shell maintainers/scripts/update.nix --argstr package skia-aseprite

This works:
> nix-shell maintainers/scripts/update.nix --argstr package aseprite

**Review by superherointj — CHANGES_REQUESTED** [review]
Please verify `skia-aseprite` update script execution, I suggest to run:

> nix-shell maintainers/scripts/update.nix --argstr package skia-aseprite

**Comment by vigress8** [comment]
> Please verify `skia-aseprite` update script execution, I suggest to run:
> 
> > nix-shell maintainers/scripts/update.nix --argstr package skia-aseprite

Reworked the script, it updates both the package and deps now, and this command works

File: pkgs/by-name/sk/skia-aseprite/update.sh [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+EOF
+
+    printf >&2 -- '----------\n\n'
+  done < <(grep -E -- "$depfilter" <<< "$third_party_deps")
+  printf '}\n'
+}
+
+update_version() {
+  local newver newrev
+  newver=$(curl -fsSL -H 'Accept: application/vnd.github+json' https://api.github.com/repos/aseprite/skia/releases/latest \
```
Comment (superherointj) [inline pkgs/by-name/sk/skia-aseprite/update.sh:46]: GitHub API has requests limits, it's good practice to include GITHUB_TOKEN to avoid issues.

One example, from another package:
> LATEST_TAG=$(curl ${GITHUB_TOKEN:+" -u \":$GITHUB_TOKEN\""} --silent https://api.github.com/repos/fluxcd/flux2/releases/latest | jq -r '.tag_name')

https://github.com/NixOS/nixpkgs/blob/5e9fc51b5be0f0c11cab93f69765f8e6a5e011fc/pkgs/by-name/fl/fluxcd/update.sh#L10

File: pkgs/by-name/sk/skia-aseprite/package.nix [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+    libjpeg
+    libpng
+    libwebp
+    libX11
+    mesa
+    zlib
+  ];
+
+  buildPhase = ''
+    runHook preBuild
```
Comment (superherointj) [inline pkgs/by-name/sk/skia-aseprite/package.nix:74]: FYI using `postBuild` allows to avoid having to add the hooks.

File: pkgs/by-name/sk/skia-aseprite/package.nix [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+  preConfigure = with depSrcs; ''
+    mkdir -p third_party/externals
+    ln -s ${angle2} third_party/externals/angle2
+    ln -s ${dng_sdk} third_party/externals/dng_sdk
+    ln -s ${piex} third_party/externals/piex
+    ln -s ${sfntly} third_party/externals/sfntly
+  '';
+
+  configurePhase = ''
+    runHook preConfigure
```
Comment (superherointj) [inline pkgs/by-name/sk/skia-aseprite/package.nix:54]: FYI: using `postConfigure` avoids having to add the hooks.

Comment (vigress8) [reply pkgs/by-name/sk/skia-aseprite/package.nix:53]: I think that these fit the definition of a `configurePhase` and `buildPhase`, so they make sense to me

**Comment by vigress8** [comment]
I've done a few other tweaks like making update.sh update the hash as well. I think this is good to go now

File: pkgs/by-name/sk/skia-aseprite/package.nix (hunk 16 above)
Comment (superherointj) [reply pkgs/by-name/sk/skia-aseprite/package.nix:53]: I don't understand what you mean.

If you use `configurePhase` you have to additionally add the hooks. If you don't, you save that, and just add what you want to be executed. My point is to give you a shorter answer, such that you have less trouble/typing. I'm not saying what you proposed is wrong. The hooks are there. It works.

To be clear, I'm not saying your code is invalid. Just that, there is an easier to do that without having to overwrite the hooks.

This is not blocking your PR. I'm informing you of a better way of doing something.

Comment (vigress8) [reply pkgs/by-name/sk/skia-aseprite/package.nix:53]: No I get what you mean.
What I mean is, a command like `gn gen` semantically makes sense to be the `configurePhase`, because it conceptually does the same thing that a `./configure` does.
Same with `ninja` - it's a build command, so putting it in `postBuild` could be confusing.
Is that clearer?

Comment (superherointj) [reply pkgs/by-name/sk/skia-aseprite/package.nix:53]: I see, you are actually replacing the `configure`. Sorry.

**Review by superherointj — APPROVED** [review]
Sorry about the delay. Been busy with so many requests.
But still, nice job. You have advanced package in many ways.

**Comment by vigress8** [comment]
Thank you for all the help!
